lib/spack/spack/test/util/package_hash.py

```diff
@@ -0,0 +1,360 @@
+    ("mfem",         "65xryd5zxarwzqlh2pojq7ykohpod4xz"),
+    ("mfem@4.0.0",   "65xryd5zxarwzqlh2pojq7ykohpod4xz"),
+    ("mfem@4.1.0",   "2j655nix3oe57iwvs2mlgx2mresk7czl"),
+    # has @when("@1.5.0:")
+    ("py-torch",     "lnwmqk4wadtlsc2badrt7foid5tl5vaw"),
+    ("py-torch@1.0", "lnwmqk4wadtlsc2badrt7foid5tl5vaw"),
+    ("py-torch@1.6", "5nwndnknxdfs5or5nrl4pecvw46xc5i2"),
+    # has a print with multiple arguments
+    ("legion",       "ba4tleyb3g5mdhhsje6t6jyitqj3yfpz"),
+])
+def test_package_hash_consistency(package_spec, expected_hash):
+    """Ensure that that package hash is consistent python version to version.
+
+    We assume these tests run across all supported Python versions in CI, and we ensure
+    consistency with recorded hashes for some well known inputs.
+
+    If this fails, then something about the way the python AST works has likely changed.
+    If Spack is running in a new python version, we might need to modify the unparser to
+    handle it. If not, then something has become inconsistent about the way we unparse
+    Python code across versions.
+
+    """
+    spec = Spec(package_spec)
+    filename = os.path.join(datadir, "%s.txt" % spec.name)
+    with open(filename) as f:
+        source = f.read()
+    h = ph.package_hash(spec, source=source)
+    assert expected_hash == h
+
+
+many_multimethods = """\
+class Pkg:
+    def foo(self):
+        print("ONE")
+
+    @when("@1.0")
+    def foo(self):
+        print("TWO")
+
+    @when("@2.0")
+    @when(sys.platform == "darwin")
+    def foo(self):
+        print("THREE")
+
+    @when("@3.0")
+    def foo(self):
+        print("FOUR")
+
+    # this one should always stay
+    @run_after("install")
+    def some_function(self):
+        print("FIVE")
+"""
+
+
+def test_multimethod_resolution():
+    # all are false but the default
+    filtered = ph.canonical_source("pkg@4.0", source=many_multimethods)
+    assert "ONE" in filtered
+    assert "TWO" not in filtered
+    assert "THREE" not in filtered
+    assert "FOUR" not in filtered
+    assert "FIVE" in filtered
+
+    # we know first @when overrides default and others are false
+    filtered = ph.canonical_source("pkg@1.0", source=many_multimethods)
+    assert "ONE" not in filtered
+    assert "TWO" in filtered
+    assert "THREE" not in filtered
+    assert "FOUR" not in filtered
+    assert "FIVE" in filtered
+
+    # we know last @when overrides default and others are false
+    filtered = ph.canonical_source("pkg@3.0", source=many_multimethods)
+    assert "ONE" not in filtered
+    assert "TWO" not in filtered
+    assert "THREE" not in filtered
+    assert "FOUR" in filtered
+    assert "FIVE" in filtered
+
+    # we don't know if default or THREE will win, include both
+    filtered = ph.canonical_source("pkg@2.0", source=many_multimethods)
+    assert "ONE" in filtered
+    assert "TWO" not in filtered
+    assert "THREE" in filtered
+    assert "FOUR" not in filtered
+    assert "FIVE" in filtered
```

Comment:
Is the explanation [in package_hash.py](https://github.com/spack/spack/pull/28156/files#diff-c0e9963b0d8dcee67b8fc64f623170a84ac7c13892f504c95106e8d1c4cdaf13R136-R178) sufficient?

The reason you need both is because you don't know until runtime whether all the `@when`'s on  `THREE` will end up being true, b/c you don't know whether you'll end up running the build on Darwin or not.  If you're on Darwin, THREE wins.  otherwise default does b/c TWO, FOUR, and FIVE are false, and the default is preserved in case nothing wins.

Reply:
Yes, I think that's sufficient. It's frustrating that we can't fully canonicalize -- I suspect that most of these packages could have `'platform=darwin'` instead of `sys.platform == 'darwin'`.

Reply:
While I was thinking about this, I realized that we should consider disallowing dynamic conditions like this in the future for another reason: it interferes with indexing the package metadata. We can deal with that later, too.